main/lsp/requests/completion.cc

```diff
@@ -92,36 +94,94 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
+
+    // if there are no lines before the method definition, we're at the top of the file.
+    if (all_lines.empty()) {
         return nullopt;
     }
 
-    string documentation = "";
-    // keep looking for previous newline locations, searching for lines with # in them.
-    while (prevLine.find('#') != prevLine.npos) {
-        documentation = absl::StrCat(prevLine.substr(prevLine.find('#') + 1, prevLine.size()), "\n", documentation);
-        int prevNewlineLoc = preDefinition.rfind('\n', lastNewlineLoc - 1);
-        // if there is no '\n', we're at the top of the file, so just return documentation.
-        if (prevNewlineLoc == preDefinition.npos) {
+    std::vector<string_view> documentation_lines;
+
+    // Iterate from the last line, to the first line
+    for (auto it = all_lines.rbegin(); it != all_lines.rend(); it++) {
+        string_view line = absl::StripAsciiWhitespace(*it);
+
+        // Short circuit when line is empty
+        if (line.empty()) {
+            break;
+        }
+
+        // Handle single-line sig block
+        else if (absl::StartsWith(line, "sig")) {
+            // Do nothing for a one-line sig block
+        }
+
+        // Handle multi-line sig block
+        else if (absl::StartsWith(line, "end")) {
+            // ASSUMPTION: We either hit the start of file, a `sig do` or an `end`
+            do
+                it++;
+            while (it != all_lines.rend()
+                   // Start of sig block
+                   && !absl::StartsWith(line, "sig do")
+                   // Invalid end keyword
+                   && !absl::StartsWith(line, "end"));
+
+            // We have either
+            // 1) Reached the start of the file
+            // 2) Found a `sig do`
+            // 3) Found an invalid end keyword
+            if (it == all_lines.rend() || absl::StartsWith(line, "end")) {
+                break;
+            } else {
+                // Reached a sig block.
+                ENFORCE(absl::StartsWith(line, "sig do"));
+
+                // Stop looking if this is a single-line block e.g `sig do; <block>; end`
+                string_view after_sig_do = line.substr(strlen("sig do"));
+                after_sig_do = absl::StripAsciiWhitespace(after_sig_do);
+                if (absl::StartsWith(after_sig_do, ";") && absl::EndsWith(after_sig_do, "end")) {
+                    break;
+                }
+                // Else, this is a valid sig block. Move on to any possible documentation.
+            }
+        }
+
+        // Handle a comment line. Do not count typing declarations.
+        else if (absl::StartsWith(line, "#") && !absl::StartsWith(line, "# typed:")) {
+            // Account for whitespace before comment e.g
+            // # abc -> "abc"
+            // #abc -> "abc"
+            int skip_after_hash = absl::StartsWith(line, "# ") ? 2 : 1;
+
+            documentation_lines.push_back(line.substr(line.find('#') + skip_after_hash, line.size()));
```

Comment:
Is the `line.size()` needed? I think it'll implicitly create a string_view from position `line.find('#') + skip_after_hash` to the end of the `string_view` if you omit the second argument.

Reply:
You're right, it's redundant. Removed in de85c6db4